### DISCUSSION TITLE
How is Hugging Face Tokenizer done ?

### DISCUSSION THREAD

**Discussion by thomasjv799** [shubham0204/Sentence-Embeddings-Android#12] [Announcements]
Hi,

I am fairly new to android and was trying to develop a RAG App. I would like to know how you are doing the tokenization properly, I tried to write logic on my own but was not good.

I saw that you are computing all variables in HFTokenizer.kt:
https://github.com/shubham0204/Sentence-Embeddings-Android/blob/ac3504e06acc45d9d7df6f40e8e3601d33560c9d/sentence_embeddings/src/main/java/com/ml/shubham0204/sentence_embeddings/HFTokenizer.kt#L14C5-L30C6

Is the above code working as you the following loading: 
https://github.com/shubham0204/Sentence-Embeddings-Android/blob/ac3504e06acc45d9d7df6f40e8e3601d33560c9d/sentence_embeddings/src/main/java/com/ml/shubham0204/sentence_embeddings/HFTokenizer.kt#L51

Is this loading the .so file in jniLibs ? If so how is this generated and does it have all functionality like python hugging face tokenizer ?

**Comment by shubham0204** [comment]
The project uses a JNI wrapper around HuggingFace [tokenizers](https://github.com/huggingface/tokenizers) (written in Rust). Check the [libs](https://github.com/shubham0204/Sentence-Embeddings-Android/tree/libs) branch of this repository to check the Rust project (see `lib.rs` for the JNI bindings). 

The `System.loadLibrary("hftokenizer")` loads the native library generated by the Rust project. This enables the Kotlin `native` functions present in `HFTokenizer.kt` to use those JNI functions present in `lib.rs.`

**Comment by thomasjv799** [comment]
Got it, thanks.
